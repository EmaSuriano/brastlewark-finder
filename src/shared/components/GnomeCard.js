import React from 'react';
import styled from 'styled-components';
import ProfessionTag from './ProfessionTag';

const GnomeContainer = styled.div`
  background: white;
  border-radius: 2px;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: start;
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
  cursor: pointer;

  &:hover {
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  }
  padding: 0 20px;
`;

const GnomePicture = styled.img.attrs({
  atl: 'Gnome profile picture',
})`
  width: 70px;
  height: 70px;
  border-radius: 50%;
  margin: auto;
  flex-shrink: 0;
  object-fit: cover;
  border: 1px ${props => props.theme.secondary} solid;
`;

const GnomeInformation = styled.span`
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  margin-left: 10px;
`;

const GnomeName = styled.h4`
  margin: 10px 0;
`;

const MAX_PROFESSIONS = 2;

const GnomeCard = ({ id, name, thumbnail, professions }) => {
  const lessProfessions =
    professions.length > MAX_PROFESSIONS
      ? professions
          .slice(0, MAX_PROFESSIONS + 1)
          .concat(`+${professions.length - MAX_PROFESSIONS}`)
      : professions;

  return (
    <GnomeContainer>
      <GnomePicture src={thumbnail} />
      <GnomeInformation>
        <GnomeName>{name}</GnomeName>
        <div>
          {lessProfessions.map(prof => (
            <ProfessionTag key={prof}>{prof}</ProfessionTag>
          ))}
        </div>
      </GnomeInformation>
    </GnomeContainer>
  );
};

export default GnomeCard;
