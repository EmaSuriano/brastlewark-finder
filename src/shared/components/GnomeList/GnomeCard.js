import React from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import Picture from '../Picture';
import Message from '../Message';
import ProfessionTag from '../ProfessionTag';

const CardGrid = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-template-rows: auto;
  grid-template-areas:
    'picture name name'
    'picture professions professions'
    'picture professions professions';
  grid-gap: 10px;
  background: white;
  border-radius: 2px;
  height: 100%;
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  box-shadow: 0 1px 3px ${props => props.theme.secondary}40,
    0 1px 2px ${props => props.theme.secondary}80;

  &:hover {
    box-shadow: 0 14px 28px ${props => props.theme.secondary}88,
      0 10px 10px ${props => props.theme.secondary}60;
  }
  padding: 0 20px;
`;

const GnomeInformation = styled.span`
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  margin-left: 10px;
`;

const GnomePicture = Picture.extend`
  grid-area: picture;
`;

const GnomeName = Message.extend`
  grid-area: name;
  margin-bottom: 0;
`;

const GnomeProfessions = styled.div`
  grid-area: professions;
`;

const StyledLink = styled(Link)`
  text-decoration: none;

  &:focus,
  &:hover,
  &:visited,
  &:link,
  &:active {
    text-decoration: none;
  }
`;

const MAX_PROFESSIONS = 2;

const GnomeCard = ({ id, name, thumbnail, professions }) => {
  const lessProfessions =
    professions.length > MAX_PROFESSIONS
      ? professions
          .slice(0, MAX_PROFESSIONS + 1)
          .concat(`+${professions.length - MAX_PROFESSIONS}`)
      : professions;

  return (
    <StyledLink to={`/gnome/${id}`}>
      <CardGrid>
        <GnomePicture size="70" src={thumbnail} />
        <GnomeName bold color="secondary">
          {name}
        </GnomeName>
        <GnomeProfessions>
          {lessProfessions.map(prof => (
            <ProfessionTag key={prof}>{prof}</ProfessionTag>
          ))}
        </GnomeProfessions>
      </CardGrid>
    </StyledLink>
  );
};

export default GnomeCard;
