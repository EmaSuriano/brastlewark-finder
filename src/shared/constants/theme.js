export default {
  background: '#4F5D75',
  light: '#FFFFFF',
  disabled: '#BFC0C0',
  accent: '#EF8354',
  secondary: '#2D3142',
};
