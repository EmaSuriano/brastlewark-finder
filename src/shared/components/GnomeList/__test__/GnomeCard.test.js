import React from 'react';
import { shallow } from 'enzyme';
import theme from 'shared/constants/theme';
import GnomeCard, {
  StyledLink,
  GnomeProfessions,
  GnomeName,
  CardGrid,
  GnomePicture,
} from '../GnomeCard';

describe('<GnomeCard />', () => {
  describe('render', () => {
    let gnomeCard;

    beforeEach(() => {
      gnomeCard = shallow(
        <GnomeCard
          id="1"
          name="James"
          thumbnail="bestPictureEver.jpg"
          professions={['chef', 'driver']}
        />,
      );
    });

    it('should render StyledLink with id', () => {
      const styledLink = gnomeCard.find('StyledLink');

      expect(styledLink.exists()).toBe(true);
      expect(styledLink.prop('to')).toBe('/gnome/1');
    });

    it('should render CardGrid with picture, name and professions', () => {
      const cardGrid = gnomeCard.find('CardGrid');

      expect(cardGrid.exists()).toBe(true);
      expect(cardGrid.find('GnomePicture').exists()).toBe(true);
      expect(cardGrid.find('GnomeName').exists()).toBe(true);
      expect(cardGrid.find('GnomeProfessions').exists()).toBe(true);
    });

    it('should render GnomePicture', () => {
      const gnomePicture = gnomeCard.find('GnomePicture');

      expect(gnomePicture.exists()).toBe(true);
      expect(gnomePicture.prop('size')).toBe(70);
      expect(gnomePicture.prop('src')).toBe('bestPictureEver.jpg');
    });

    describe('GnomeProfessions with Professions Tag', () => {
      it('should render the same amount if professions are less than MAX_PROFESSIONS', () => {
        const gnomeProfessions = gnomeCard.find('GnomeProfessions');

        expect(gnomeProfessions.exists()).toBe(true);
        expect(gnomeProfessions.find('ProfessionTag').length).toBe(2);
      });

      it('should render the upn to MAX_PROFESSIONS amount if professions are greater than MAX_PROFESSIONS', () => {
        gnomeCard.setProps({
          professions: ['chef', 'driver', 'runner', 'painter', 'skater'],
        });

        const gnomeProfessions = gnomeCard.find('GnomeProfessions');

        expect(gnomeProfessions.exists()).toBe(true);
        expect(gnomeProfessions.find('ProfessionTag').length).toBe(4);
      });
    });
  });
});

describe('<StyledLink />', () => {
  it('render', () => {
    expect(shallow(<StyledLink to="/" />)).toMatchSnapshot();
  });
});

describe('<GnomeProfessions />', () => {
  it('render', () => {
    expect(shallow(<GnomeProfessions />)).toMatchSnapshot();
  });
});

describe('<GnomeName />', () => {
  it('render', () => {
    expect(shallow(<GnomeName />)).toMatchSnapshot();
  });
});

describe('<GnomePicture />', () => {
  it('render', () => {
    expect(shallow(<GnomePicture size={1} />)).toMatchSnapshot();
  });
});

describe('<CardGrid />', () => {
  it('render', () => {
    expect(shallow(<CardGrid theme={theme} />)).toMatchSnapshot();
  });
});
