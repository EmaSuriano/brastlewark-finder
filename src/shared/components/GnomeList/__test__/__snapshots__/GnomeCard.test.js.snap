// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardGrid /> render 1`] = `
.c0 {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-template-rows: auto;
  grid-template-areas: 'picture name name' 'picture professions professions' 'picture professions professions';
  grid-gap: 10px;
  background: white;
  border-radius: 2px;
  height: 100%;
  -webkit-transition: all 0.3s cubic-bezier(0.25,0.8,0.25,1);
  transition: all 0.3s cubic-bezier(0.25,0.8,0.25,1);
  box-shadow: 0 1px 3px #2D314240, 0 1px 2px #2D314280;
  padding: 0 20px;
}

.c0:hover {
  box-shadow: 0 14px 28px #2D314288, 0 10px 10px #2D314260;
}

<div
  className="c0"
/>
`;

exports[`<GnomeName /> render 1`] = `
.c0 {
  font-weight: inherit;
  font-size: inherit;
  grid-area: name;
  margin-bottom: 0;
}

<p
  className="c0"
  color="light"
  fontSize="inherit"
/>
`;

exports[`<GnomePicture /> render 1`] = `
.c0 {
  width: 1px;
  height: 1px;
  border-radius: 50%;
  margin: auto;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  object-fit: cover;
  border: 1px solid;
  grid-area: picture;
}

<img
  atl="Gnome profile picture"
  className="c0"
  size={1}
/>
`;

exports[`<GnomeProfessions /> render 1`] = `
.c0 {
  grid-area: professions;
}

<div
  className="c0"
/>
`;

exports[`<StyledLink /> render 1`] = `
.c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:focus,
.c0:hover,
.c0:visited,
.c0:link,
.c0:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Link
  className="c0"
  replace={false}
  to="/"
/>
`;
